### PR TITLE
Headless prompt sdk

### PR DESCRIPTION
**Pull request by nicolewhite**
Usage will look like:

```ts
import { AutoblocksPromptManager } from '@autoblocks/client/prompts';

const manager = new AutoblocksPromptManager({
  id: 'prompt-id',
  version: {
    major: '1',
    minor: '0',
  },
});
```

or

```ts
const manager = new AutoblocksPromptManager({
  id: 'prompt-id',
  version: {
    major: '1',
    minor: 'latest',
  },
});
```

or

```ts
const manager = new AutoblocksPromptManager({
  id: 'prompt-id',
  version: {
    major: '1',
    minor: [
      {
        version: 'latest',
        weight: 10,
      },
      {
        version: '0',
        weight: 90,
      },
    ]
  },
});
```

Then:

```ts
const response = await manager.exec(async ({ prompt }) => {
  const tracer = new AutoblocksTracer( {
    traceId: crypto.randomUUID(),
  });
  const params: ChatCompletionCreateParamsNonStreaming = {
    model: prompt.params.model,
    temperature: prompt.params.temperature,
    messages: [
      {
        role: 'system',
        content: prompt.render({ template: 'system', params: {...} }),
      },
      {
        role: 'user',
        content: prompt.render({ template: 'user', params: {...} }),
      },
    ],
  };
  await tracer.sendEvent('ai.request', { properties: params });
  const response = await openai.chat.completions.create(params);
  await tracer.sendEvent(
    'ai.response',
    {
      properties: {
        response,
      },
      promptTracking: prompt.track(),
    },
  );
  return response;
});
```

The motivation behind `exec` is that we need to "freeze" the current prompt so that if the `latest` version changes between when `render` is called and when `track` is called, we record the correct version in the tracking info.

Also wasn't able to get the types working for specifying multiple versions on one manager instance, since I don't think there's any way to get compile-time type safety. For example:

```ts
const manager = new AutoblocksPromptManager({
  subscriptions: [
    {
      id: 'prompt-id-1',
      version: {
        major: '1',
        minor: '0',
      },
    },
    {
      id: 'prompt-id-2',
      version: {
        major: '1',
        minor: '0',
      },
    },
  ],
});
```

We can have type safety for the subscription list (in that we can verify it belongs to known prompt IDs and their versions) but AFAIK there is no way to have type safety to verify subsequently-chosen prompt IDs in `exec` were in your subscription list, since the subscription list is only known at runtime. I think the only way we could do something like that would involve having users specify their subscriptions in another place, like a YAML file that the autogeneration CLI could see, and only generate types for the subscribed-to-prompts and not every prompt.

So long story short if you have multiple prompts, you'll have to do something like:

```ts
const managerA = new AutoblocksPromptManager(...);
const managerB = new AutoblocksPromptManager(...);

const allManagers = [
  managerA,
  managerB,
];

// On app startup
await Promise.all(allManagers.map(m => m.init());

// On app shutdown
allManagers.forEach(m => m.close());
```

### CODE REVIEW

File: src/prompts/headless/manager.ts [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+    }
+    this.apiKey = apiKey;
+
+    if (
+      this.minorVersionsToRequest.some(
+        (version) => version === HEADLESS_PROMPT_LATEST_VERSION,
+      )
+    ) {
+      const refreshInterval = args.refreshInterval || { seconds: 10 };
+      console.log(
```
Comment (adamnolte) [inline src/prompts/headless/manager.ts:81]: I think a future enhancement that would probably be nice is the ability to pass in a logger

File: src/prompts/headless/manager.ts [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+        `[${this.id}@v${this.majorVersion}] Refreshing latest prompt every ${refreshInterval.seconds} seconds`,
+      );
+      this.refreshIntervalTimer = setInterval(
+        this.refreshLatest,
+        convertTimeDeltaToMilliSeconds(refreshInterval),
+      );
+    }
+
+    this.refreshTimeout = args.refreshTimeout || { seconds: 5 };
+    this.initTimeout = args.initTimeout || { seconds: 5 };
```
Comment (adamnolte) [inline src/prompts/headless/manager.ts:91]: I think we may need a bit more of a generous default for folks that aren't near us-east-1

Comment (adamnolte) [reply src/prompts/headless/manager.ts:91]: maybe 30 seconds

**Comment by adamnolte** [comment]
Spent some time thinking through the exec method and naming of render.

i think both make sense. The alternative I was thinking about for exec was returning the prompt instead of passing a function as an arg, but there is risk then that folks continue to use that prompt so it doesn't gets refreshed with the latest or they make it a singleton.

What might be nice in the future is if we figure out how to automatically track the llm calls inside of exec. But maybe if we do that we should just do it globally and not only inside of exec.

I think the flexibility of how it is now, is nice.

The naming `render` for the prompts did give me pause at first, but I think it probably is the most appropriate name. The other option I thought of was `make`.

**Comment by nicolewhite** [comment]
yeah agreed on all points

> The alternative I was thinking about for exec was returning the prompt instead of passing a function as an arg

i had considered this as well but didn't go that route for the same reason you described

yeah for some reason `build` was starting to sound weird but is potentially a more accurate term than `render` so could switch back to that

**Comment by nicolewhite** [comment]
> What might be nice in the future is if we figure out how to automatically track the llm calls inside of exec

yeah i haven't tried it yet but i was thinking we could do something where you pass your `openai` instance as an argument to `AutoblocksPromptManager`, we wrap it, then include it in `exec`. so:

```ts
const openai = new OpenAI();
const manager = new AutoblocksPromptManager({
  id: 'prompt-id',
  version: {
    major: '1',
    minor: '0',
  },
  openai,
});
```

Then:

```ts
const response = await manager.exec(async ({ prompt, openai, tracer }) => {
  // Make params
  const params = ...
  // Call openai w/ wrapped openai instance
  const response = await openai.chat.completions.create(params);
  // Send custom event
  await tracer.sendEvent(...);
});
```

`openai` would be a wrapped openai instance and `tracer` would be the tracer instance used by the openai wrapper so that you could log custom events w/ the same trace ID

**Review by adamnolte — APPROVED** [review]
